src/lib/coda_base/sparse_ledger.ml

```diff
@@ -122,49 +130,157 @@ let sub_account_creation_fee action (amount : Currency.Amount.t) =
         sub amount (of_fee Coda_compile_config.account_creation_fee))
   else amount
 
-let apply_user_command_exn t ({sender; payload; signature= _} : User_command.t)
-    =
-  let sender_idx = find_index_exn t (Public_key.compress sender) in
-  let sender_account = get_exn t sender_idx in
-  let nonce = User_command.Payload.nonce payload in
-  let fee = User_command.Payload.fee payload in
-  assert (Account.Nonce.equal sender_account.nonce nonce) ;
+let apply_user_command_exn t
+    ({signer; payload; signature= _} as user_command : User_command.t) =
   let open Currency in
-  let sender_account =
-    { sender_account with
-      nonce= Account.Nonce.succ sender_account.nonce
-    ; receipt_chain_hash=
-        Receipt.Chain_hash.cons payload sender_account.receipt_chain_hash
-    ; balance=
-        Balance.sub_amount sender_account.balance (Amount.of_fee fee)
-        |> Option.value_exn ?here:None ?error:None ?message:None }
+  let signer_pk = Public_key.compress signer in
+  (* TODO: Put actual value here. See issue #4036. *)
+  let current_global_slot = Coda_numbers.Global_slot.zero in
+  (* Fee-payer information *)
+  let fee_token = User_command.fee_token user_command in
+  let fee_payer = User_command.fee_payer user_command in
+  let nonce = User_command.nonce user_command in
+  let () =
+    (* TODO: Enable multi-sig. *)
+    assert (
+      Public_key.Compressed.equal (Account_id.public_key fee_payer) signer_pk
+    )
   in
-  match User_command.Payload.body payload with
-  | Stake_delegation (Set_delegate {new_delegate}) ->
-      set_exn t sender_idx {sender_account with delegate= new_delegate}
-  | Payment {amount; receiver} ->
-      let t =
-        set_exn t sender_idx
-          { sender_account with
+  let () = assert (Token_id.equal fee_token Token_id.default) in
```

Comment:
Minor nit, but we typically prefer `;` for sequencing units rather than the traditional SML style `let () = ... in`.